### ISSUE TITLE
Issue while reading data directly from S3

### ISSUE THREAD

**Issue by renato2099** [Unidata/netcdf-c#1497]
To report a non-security related issue, please provide:

* the version of the software with which you are encountering an issue
current master in GitHub

* environmental information (i.e. Operating System, compiler info, java version, python version, etc.)
Ubuntu 16.04
* a description of the issue with the steps needed to reproduce it
Trying to read a publicly available file from Amazon S3

If you have a general question about the software, please view our [Suggested Support Process](https://www.unidata.ucar.edu/support/#process).

Hi Unidata team,

I am unsure if this is the right place the ask, but it definitely seems the most active. So please forgive me if this is not the correct place.
I am trying to read a public file from S3 through `nc_open()` but I am getting the following error:

`syntax error, unexpected WORD_WORD, expecting SCAN_ATTR or SCAN_DATASET or SCAN_ERROR
context: <?xml^ version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>3B-HHR.HDF5.dds</Key><RequestId>C4E7BACB0B8C4258</RequestId><HostId>tROyJy7+a8i/GBuZZQFQXtl95WaKYR4jsK0EBTunVs5Cdr39oswo8nuGYwDBXIkSTHIJ2iJIju0=</HostId></Error>
`
I came across [this](https://github.com/Unidata/netcdf-c/pull/1267/files)  so I assume this is possible, and I am just doing something wrong? Thanks in advanced!

**Comment by WardF** [comment]
Can you provide a little more info about how you are trying to open the file? What are you passing to `nc_open()`?   @DennisHeimbigner might be able to provide some insight as well.

**Comment by DennisHeimbigner** [comment]
Yes, it is possible since we tested it against some similar S3 files.
We will need to know the exact url that you used.

**Comment by renato2099** [comment]
Thanks for the prompt reply!

https://tpch-rmm-fk.s3.eu-central-1.amazonaws.com/3B.nc4

Currently I am using something as simple as this:

`(retval = nc_open(argv[1], NC_NOWRITE, &ncid))`

**Comment by DennisHeimbigner** [comment]
You need a marker to tell the library to use byte ranges.
Change the url to this and try again:
"https://tpch-rmm-fk.s3.eu-central-1.amazonaws.com/3B.nc4#mode=bytes"

**Comment by renato2099** [comment]
Thank you very much @DennisHeimbigner ! that worked like a charm.
I still have some questions though:
1) When should one prefer using `nc_http_open` vs `nc_open` api? are they the same or is one subsumed by the other?
2) If I were to track the total number of requests done to S3, which files should I look into?
3) Can I make the same assumptions when reading files from S3 as I would when reading files from a local file system? e.g., in terms of data access, are variables accessed only once contiguously if I use `nc_get_var_...`? I guess at least I will do 1 request for file metadata (is the size of this metadata always the same across nc4 files?) and then 1 request per variable, right?

**Comment by DennisHeimbigner** [comment]
>1. When should one prefer using nc_http_open vs nc_open api? are they the same or is one subsumed by the other?

The public API for the netcdf library is defined by the union of
the files in include that are of the form netcdf_XXX.h. The
include directory however does contain other, non-public, .h
files; nchttp.h is one such.

Bottom line: since nc_http_open is not public, you should always
use nc_open, which is public.


>2. If I were to track the total number of requests done to S3, which files should I look into?

Not sure I understand the question. Are you asking where to insert code
to track # reads?


>3. Can I make the same assumptions when reading files from S3 as I would when reading files from a local file system? e.g., in terms of data access, are variables accessed only once contiguously if I use nc_get_var_...?

Yes, as far as I know. There is possibly a complication with respect to
chunking. It is possible that when reading what seems to be a contiguous
piece of data, that the HDF5 library will need to read pieces from multiple
chunks and assemble the pieces in memory into one large piece.

> I guess at least I will do 1 request for file metadata (is the size of this metadata always the same across nc4 files?)

Metadata size can vary depending on a whole variety of factors, to the number
of reads to read the metadata (for a variable, I assume) might be different
for variables of, for example, different types and rank.

> and then 1 request per variable, right?
Yes, except as noted above.
